File: web/src/api/service.ts
import { api } from './client';

export interface ServiceStatus {
	service_running: boolean;
	last_heartbeat: string | null;
	last_task_execution: string | null;
	error_count: number;
	last_error: string | null;
	updated_at: string | null;
}

export interface ServiceStartResponse {
	success: boolean;
	message: string;
	service_running: boolean;
}

export interface ServiceStopResponse {
	success: boolean;
	message: string;
	service_running: boolean;
}

export interface TaskExecution {
	id: number;
	task_name: string;
	executed_at: string;
	status: 'success' | 'failed' | 'skipped';
	duration_seconds: number;
	details: Record<string, any> | null;
}

export interface TaskHistory {
	tasks: TaskExecution[];
	total: number;
}

export interface ServiceLog {
	id: number;
	level: 'DEBUG' | 'INFO' | 'WARNING' | 'ERROR' | 'CRITICAL';
	module: string;
	message: string;
	context: Record<string, any> | null;
	timestamp: string;
}

export interface ServiceLogs {
	logs: ServiceLog[];
	total: number;
	limit: number;
}

export async function startService(): Promise<ServiceStartResponse> {
	const { data } = await api.post<ServiceStartResponse>('/user/service/start');
	return data;
}

export async function stopService(): Promise<ServiceStopResponse> {
	const { data } = await api.post<ServiceStopResponse>('/user/service/stop');
	return data;
}

export async function getServiceStatus(): Promise<ServiceStatus> {
	const { data } = await api.get<ServiceStatus>('/user/service/status');
	return data;
}

export async function getTaskHistory(params?: {
	task_name?: string;
	status?: 'success' | 'failed' | 'skipped';
	limit?: number;
}): Promise<TaskHistory> {
	const { data } = await api.get<TaskHistory>('/user/service/tasks', { params });
	return data;
}

export async function getServiceLogs(params?: {
	level?: 'DEBUG' | 'INFO' | 'WARNING' | 'ERROR' | 'CRITICAL';
	module?: string;
	hours?: number;
	limit?: number;
}): Promise<ServiceLogs> {
	const { data } = await api.get<ServiceLogs>('/user/service/logs', { params });
	return data;
}

File: server/app/schemas/service.py
"""API schemas for service management"""

from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, Field


class ServiceStatusResponse(BaseModel):
    """Service status response"""

    service_running: bool
    last_heartbeat: datetime | None = None
    last_task_execution: datetime | None = None
    error_count: int = 0
    last_error: str | None = None
    updated_at: datetime | None = None


class ServiceStartResponse(BaseModel):
    """Response for service start operation"""

    success: bool
    message: str
    service_running: bool


class ServiceStopResponse(BaseModel):
    """Response for service stop operation"""

    success: bool
    message: str
    service_running: bool


class TaskExecutionResponse(BaseModel):
    """Task execution history item"""

    id: int
    task_name: str
    executed_at: datetime
    status: Literal["success", "failed", "skipped"]
    duration_seconds: float
    details: dict[str, Any] | None = None


class TaskHistoryResponse(BaseModel):
    """Task execution history response"""

    tasks: list[TaskExecutionResponse]
    total: int


class ServiceLogResponse(BaseModel):
    """Service log entry"""

    id: int
    level: Literal["DEBUG", "INFO", "WARNING", "ERROR", "CRITICAL"]
    module: str
    message: str
    context: dict[str, Any] | None = None
    timestamp: datetime


class ServiceLogsResponse(BaseModel):
    """Service logs response"""

    logs: list[ServiceLogResponse]
    total: int
    limit: int = Field(default=100, description="Number of logs returned")


# Individual Service Management Schemas


class IndividualServiceStatus(BaseModel):
    """Individual service status"""

    task_name: str
    is_running: bool
    started_at: datetime | None = None
    last_execution_at: datetime | None = None
    next_execution_at: datetime | None = None
    process_id: int | None = None
    schedule_enabled: bool = True
    last_execution_status: Literal["success", "failed", "skipped", "running"] | None = None
    last_execution_duration: float | None = None
    last_execution_details: dict[str, Any] | None = None


class IndividualServicesStatusResponse(BaseModel):
    """Response for individual services status"""

    services: dict[str, IndividualServiceStatus]


class StartIndividualServiceRequest(BaseModel):
    """Request to start individual service"""

    task_name: str = Field(
        ...,
        description=(
            "Task name: premarket_retry, sell_monitor, position_monitor, buy_orders, eod_cleanup"
        ),
    )


class StartIndividualServiceResponse(BaseModel):
    """Response for starting individual service"""

    success: bool
    message: str


class StopIndividualServiceRequest(BaseModel):
    """Request to stop individual service"""

    task_name: str = Field(..., description="Task name to stop")


class StopIndividualServiceResponse(BaseModel):
    """Response for stopping individual service"""

    success: bool
    message: str


class RunOnceRequest(BaseModel):
    """Request to run task once"""

    task_name: str = Field(..., description="Task name to run")
    execution_type: Literal["run_once", "manual"] = Field(
        default="run_once", description="Execution type"
    )


class RunOnceResponse(BaseModel):
    """Response for run once execution"""

    success: bool
    message: str
    execution_id: int | None = None
    has_conflict: bool = False
    conflict_message: str | None = None


# Service Schedule Management Schemas (Admin)


class ServiceScheduleResponse(BaseModel):
    """Service schedule response"""

    id: int
    task_name: str
    schedule_time: str  # HH:MM format
    enabled: bool
    is_hourly: bool
    is_continuous: bool
    end_time: str | None = None  # HH:MM format
    schedule_type: str = "daily"  # "daily" or "once"
    description: str | None = None
    updated_by: int | None = None
    updated_at: datetime
    next_execution_at: datetime | None = None


class ServiceSchedulesResponse(BaseModel):
    """List of service schedules"""

    schedules: list[ServiceScheduleResponse]


class UpdateServiceScheduleRequest(BaseModel):
    """Request to update service schedule"""

    schedule_time: str = Field(..., description="Time in HH:MM format (IST)")
    enabled: bool = Field(default=True)
    is_hourly: bool = Field(default=False)
    is_continuous: bool = Field(default=False)
    end_time: str | None = Field(
        None, description="End time in HH:MM format (for continuous tasks)"
    )
    schedule_type: str = Field(
        default="daily", description="'daily' runs every day, 'once' runs once and stops"
    )
    description: str | None = None


class UpdateServiceScheduleResponse(BaseModel):
    """Response for updating service schedule"""

    success: bool
    message: str
    schedule: ServiceScheduleResponse
    requires_restart: bool = Field(
        default=True, description="Whether unified service needs to restart to apply changes"
    )
